Allow input of purchases via a file, 
containing data in the JSON format specified below.

Parse the purchase data and calculate the following statistics:

- total volume of spend
- average purchase value
- maximum purchase value
- median purchase value
- number of unique products purchased
- Hint: purchase value will need to be computed

Print results to STDOUT in a JSON format

Run on either Linux or Mac OS X

File loaded successfully!


In [122]:
import os
import json

# Get complete current working directory (to ensure the script works regardless of where it is being executed)
## Assumption: file will be available within the same directory
current_directory = os.getcwd()
file_path = os.path.join(current_directory, 'purchases_v1.json')


try:
    with open(file_path, 'r') as file:
        data = json.load(file)
    print("File loaded successfully!")
except FileNotFoundError:
    print(f"file {file_path} was not found.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")


total_volume_of_spend = 0
unique_purchases = []
purchase_data = {}  # Dictionary to store purchase_id as keys and purchase_total as values

#Using set instead of list as it provides a set of unique elements by default
unique_products = set()

for purchase in data:
    value_per_purchase = 0
    for item in purchase['items']:
        # Store quantity, purchase and price
        quantity = int(item.get('quantity', 0))  # Default to 0 if quantity is missing
        price = float(item.get('price', 0))  # Default to 0 if price is missing
        total_volume_of_spend += price * quantity
        value_per_purchase += price * quantity
        unique_products.add(item['product_name']) # Automatically check unique product name
        
    # Store unique purchases, Assumption: purchase id might not be unique
    if purchase['purchase_id'] not in data:
        purchase_data[purchase['purchase_id']] = value_per_purchase
        # As a dict, store purchase_id and price

# Calculate average
if len(purchase_data) > 0:
    avg_purchase_value = total_volume_of_spend / len(purchase_data)
else:
    avg_purchase_value = 0  # Handle division by zero


# Calculate max value if the dict is non null
max_purchase_value = max(purchase_data.values(), default = 0)

#maximum_purchase_value = max()

# Calculate the median purchase value

'''
Example

# [1, 2, 3, 4, 5, 6]
# (3+4)/2
# (list[2] + list[3]) / 2
# ((len(list) - 1)/2) + len(list)/2

'''

if len(purchase_data) > 0:
    # First need to get sorted values in a list
    sorted_purchase_values = sorted(purchase_data.values())
    length_purchase_values = len(sorted_purchase_values)
    index = (length_purchase_values - 1) // 2

    if (length_purchase_values % 2 == 0):
        median = (sorted_purchase_values[index] + sorted_purchase_values[index + 1])/2
    else:
        median = sorted_purchase_values[index]
else:
    print("No data")
    median = 0


# Get the number of unique products purchased
number_of_unique_products = len(unique_products)


# Print final results
results = {
    "Total Volume of Spend": total_volume_of_spend,
    "Average purchase value": avg_purchase_value,
    "Maximum purchase value": max_purchase_value,
    "Median purchase value": median,
    "Number of unique products purchased": number_of_unique_products
}

# Print the results in JSON format
print(json.dumps(results, indent=4))

File loaded successfully!
{
    "Total Volume of Spend": 101975.0,
    "Average purchase value": 2039.5,
    "Maximum purchase value": 4721.0,
    "Median purchase value": 2043.5,
    "Number of unique products purchased": 88
}


- Total volume of spend: $101,975.00
- Average purchase value: $2,039.50
- Maximum purchase value: $4,721.00
- Median purchase value: $2,043.50
- Number of unique products purchased: 88